File: Khanc/src/Lex/lexer.c
#include "Lex/lexer.h"
#include "error.h"
#include <ctype.h>
#include <stdlib.h>
#include <string.h>
#include "Token.h"
#include <stdio.h>

void initLexer(Lexer *lexer, const char *input) {
    lexer->input = input;
    lexer->cur = 0;
    lexer->line = 1;
    lexer->column = 1;
}

Token nextToken(Lexer *lexer) {
    char c;

    while ((c = lexer->input[lexer->cur]) != '\0') {
        // Beyaz boşlukları atla
        if (isspace(c)) {
            if (c == '\n') {
                lexer->line++;
                lexer->column = 1;
            } else {
                lexer->column++;
            }
            lexer->cur++;
            continue;
        }

        switch (c) {
            case '+':
                lexer->cur++;
                lexer->column++;
                return createToken(PLUS, "+");
            case '-':
                lexer->cur++;
                lexer->column++;
                return createToken(MINUS, "-");
            case '*':
                lexer->cur++;
                lexer->column++;
                return createToken(MULTIPLY, "*");
            case '/':
                lexer->cur++;
                lexer->column++;
                return createToken(DIVIDE, "/");
            case ';':
                lexer->cur++;
                lexer->column++;
                return createToken(SEMICOLON, ";");
            case '=':
                lexer->cur++;
                lexer->column++;
                return createToken(EQUALS, "=");
            case '(':
                lexer->cur++;
                lexer->column++;
                return createToken(LPAREN, "(");
            case ')':
                lexer->cur++;
                lexer->column++;
                return createToken(RPAREN, ")");
            case ',':
                lexer->cur++;
                lexer->column++;
                return createToken(COMMA, ",");
            case ':':
                if (lexer->input[lexer->cur + 1] == '=') {
                    lexer->cur += 2;
                    lexer->column += 2;
                    return createToken(TTEQUALS, ":=");
                } else {
                    lexer->cur++;
                    lexer->column++;
                    return createToken(TDOT, ":");
                }
            case '{':
                lexer->cur++;
                lexer->column++;
                return createToken(LBRACE, "{");
            case '}':
                lexer->cur++;
                lexer->column++;
                return createToken(RBRACE, "}");
            default:
                if (isdigit(c)) {
                    int start = lexer->cur;
                    while (isdigit(lexer->input[lexer->cur])) {
                        lexer->cur++;
                        lexer->column++;
                    }

                    int length = lexer->cur - start;
                    char *value = (char *)malloc(length + 1);
                    strncpy(value, &lexer->input[start], length);
                    value[length] = '\0';

                    Token token = createToken(INT, value);
                    free(value);

                    return token;
                }
                else if (isalnum(c) || c == '_') {
                    int start = lexer->cur;
                    while (isalnum(lexer->input[lexer->cur]) || lexer->input[lexer->cur] == '_') {
                        lexer->cur++;
                        lexer->column++;
                    }

                    int length = lexer->cur - start;
                    char *value = (char *)malloc(length + 1);
                    strncpy(value, &lexer->input[start], length);
                    value[length] = '\0';

                    // Anahtar kelimeleri kontrol et
                    if (strcmp(value, "let") == 0) {
                        Token token = createToken(LET, value);
                        free(value);
                        return token;
                    }
                    else if (strcmp(value, "elif") == 0) {
                        Token token = createToken(ELSE_IF, value);
                        free(value);
                        return token;
                    }
                    else if (strcmp(value, "if") == 0) {
                        Token token = createToken(IF, value);
                        free(value);
                        return token;
                    }
                    else if (strcmp(value, "else") == 0) {
                        Token token = createToken(ELSE, value);
                        free(value);
                        return token;
                    }
                    else if (strcmp(value, "while") == 0) {
                        Token token = createToken(WHILE, value);
                        free(value);
                        return token;
                    }
                    else {
                        Token token = createToken(IDENTIFIER, value);
                        free(value);
                        return token;
                    }
                }
                else {
                    error(LEXICAL_ERROR, lexer->line, lexer->column, "Unexpected character '%c'", lexer->input[lexer->cur]);
                    lexer->cur++;
                    lexer->column++;
                }
        }
    }
    checkError(lexer->input);
    return createToken(END_OF_FILE, "EOF");
}

File: Khanpy/Lexer/Lexer.py
from Token.Token import TokenType, Token

class Lexer:
    def __init__(self, text):
        self.text = text
        self.pos = 0
        self.cur = self.text[self.pos] if self.text else None

    def advance(self):
        self.pos += 1
        if self.pos < len(self.text):
            self.cur = self.text[self.pos]
        else:
            self.cur = None  # EOF durumu

    def nextToken(self):
        while self.cur is not None:  # Boşlukları atla
            if self.cur.isspace():
                self.advance()
                continue

            # İki karakterli operatörleri kontrol et
            if self.cur == '=':
                self.advance()
                if self.cur == '=':
                    self.advance()
                    return Token(TokenType.EQUALS_EQUALS, '==')
                return Token(TokenType.EQUALS, '=')
            if self.cur == '!':
                self.advance()
                if self.cur == '=':
                    self.advance()
                    return Token(TokenType.NOT_EQUALS, '!=')
                raise Exception(f'Invalid token: {self.cur}')
            if self.cur == '>':
                self.advance()
                if self.cur == '=':
                    self.advance()
                    return Token(TokenType.GTE, '>=')
                return Token(TokenType.GT, '>')
            if self.cur == '<':
                self.advance()
                if self.cur == '=':
                    self.advance()
                    return Token(TokenType.LTE, '<=')
                return Token(TokenType.LT, '<')
            match self.cur:
                case '+':
                    self.advance()
                    return Token(TokenType.PLUS, '+')
                case '-':
                    self.advance()
                    if self.cur == '>':
                        self.advance()
                        return Token(TokenType.TO, '->')
                    return Token(TokenType.MINUS, '-')
                case '*':
                    self.advance()
                    return Token(TokenType.MUL, '*')
                case '/':
                    self.advance()
                    return Token(TokenType.DIV, '/')
                case ';':
                    self.advance()
                    return Token(TokenType.SEMICOLON, ';')
                case ',':
                    self.advance()
                    return Token(TokenType.COMMA, ',')
                case '(':
                    self.advance()
                    return Token(TokenType.LPAREN, '(')
                case ')':
                    self.advance()
                    return Token(TokenType.RPAREN, ')')
                case '{':
                    self.advance()
                    return Token(TokenType.LBRACE, '{')
                case '}':
                    self.advance()
                    return Token(TokenType.RBRACE, '}')
                case '.':
                    self.advance()
                    return Token(TokenType.DOT, '.')
                case '#':
                    self.advance()
                    return Token(TokenType.HASHTAG, '#')
                case '\"':
                    self.advance()
                    value = ''
                    while self.cur is not None and self.cur != '\"':
                        value += self.cur
                        self.advance()
                    self.advance()
                    return Token(TokenType.IDENTIFIER, f"\"{value}\"")
                case ':':
                    self.advance()
                    if self.cur == '=':
                        self.advance()
                        return Token(TokenType.AUTO, ':=')
                    else:
                        return Token(TokenType.TTDOT, ':')
                case _:
                    if self.cur.isdigit():
                        value = ''
                        while self.cur is not None and self.cur.isdigit():
                            value += self.cur
                            self.advance()
                        return Token(TokenType.INT, value)
                    elif self.cur.isalpha():
                        value = ''
                        while self.cur is not None and self.cur.isalpha():
                            value += self.cur
                            self.advance()
                        match value:
                            case 'if':
                                return Token(TokenType.IF, value)
                            case 'else':
                                return Token(TokenType.ELSE, value)
                            case 'elif':
                                return Token(TokenType.ELIF, value)
                            case 'let':
                                return Token(TokenType.LET, value)
                            case 'import':
                                return Token(TokenType.IMPORT, value)
                            case 'include':
                                return Token(TokenType.INCLUDE, value)
                            case 'func':
                                return Token(TokenType.FUNC, value)
                            case 'return':
                                return Token(TokenType.RETURN, value)
                            case _:
                                return Token(TokenType.IDENTIFIER, value)
                    else:
                        raise Exception(f'Invalid token: {self.cur}')
        
        return Token(TokenType.EOF, None)  # EOF token'ını döndür
